cdap-standalone/src/main/java/co/cask/cdap/TrackerAppCreationService.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap;
+
+import co.cask.cdap.api.artifact.ArtifactScope;
+import co.cask.cdap.api.artifact.ArtifactVersion;
+import co.cask.cdap.client.ApplicationClient;
+import co.cask.cdap.client.ArtifactClient;
+import co.cask.cdap.client.ProgramClient;
+import co.cask.cdap.client.config.ClientConfig;
+import co.cask.cdap.client.config.ConnectionConfig;
+import co.cask.cdap.common.ArtifactNotFoundException;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.conf.KafkaConstants;
+import co.cask.cdap.common.utils.Tasks;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.artifact.AppRequest;
+import co.cask.cdap.proto.artifact.ArtifactSummary;
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Create Tracker App in Default Namespace and start its Programs during Standalone startup.
+ */
+public class TrackerAppCreationService extends AbstractExecutionThreadService {
+
+  private final CConfiguration cConf;
+  private final ClientConfig clientConfig;
+  private ArtifactSummary artifactSummary;
+
+  @Inject
+  public TrackerAppCreationService(CConfiguration cConf) {
+    this.cConf = cConf;
+    String address = cConf.get(Constants.Router.ADDRESS);
+    int routerPort = cConf.getInt(Constants.Router.ROUTER_PORT);
+    this.clientConfig = ClientConfig.builder().setConnectionConfig(
+      ConnectionConfig.builder().setHostname(address).setPort(routerPort).build()).build();
+  }
+
+  @Override
+  protected void run() throws Exception {
+    // Wait until tracker artifact is available
+    final ArtifactClient client = new ArtifactClient(clientConfig);
```

Comment:
Why use `ArtifactClient`?? Everything is within the same JVM, doesn't make sense to use HTTP. Can't you just get a `ArtifactRepository` injected?